File: backend/src/main/java/com/NBE3_4_2_Team4/global/security/jwt/JwtManager.java
package com.NBE3_4_2_Team4.global.security.jwt;

import com.NBE3_4_2_Team4.domain.member.member.entity.Member;
import com.NBE3_4_2_Team4.domain.member.member.repository.MemberRepository;
import com.NBE3_4_2_Team4.global.security.user.tempUserBeforeSignUp.TempUserBeforeSignUp;
import com.NBE3_4_2_Team4.standard.constants.AuthConstants;
import io.jsonwebtoken.*;
import io.jsonwebtoken.security.Keys;
import lombok.Getter;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

import javax.crypto.SecretKey;
import java.util.Base64;
import java.util.Date;
import java.util.HashMap;
import java.util.Map;
@Slf4j
@Component
public class JwtManager {
    @Getter
    private final int accessTokenValidMinute;
    @Getter
    private final int refreshTokenValidHour;
    private final SecretKey key;
    private final MemberRepository memberRepository;

    public JwtManager(
            @Value("${custom.jwt.secretKey:key}") String jwtSecretKey,
            @Value("${custom.jwt.accessToken.validMinute:30}") int accessTokenValidMinute,
            @Value("${custom.jwt.refreshToken.validHour:24}") int refreshTokenValidHour,
            MemberRepository memberRepository
            ){
        try {
            byte[] keyBytes = Base64.getDecoder().decode(jwtSecretKey);
            this.key = Keys.hmacShaKeyFor(keyBytes);
            this.accessTokenValidMinute = accessTokenValidMinute;
            this.refreshTokenValidHour = refreshTokenValidHour;
            this.memberRepository = memberRepository;
        }catch (IllegalArgumentException e){
            throw new RuntimeException("키 값은 Base64로 디코딩 가능한 값이어야 합니다. yml 관련 파일을 확인해보세요.");
        }
    }

    public String generateAccessToken(Member member) {
        return Jwts.builder()
                .claim(AuthConstants.ID, member.getId())
                .claim(AuthConstants.USERNAME, member.getUsername())
                .claim(AuthConstants.NICKNAME, member.getNickname())
                .claim(AuthConstants.ROLE, member.getRole().name())
                .claim(AuthConstants.OAUTH2_PROVIDER, member.getOAuth2Provider().name())
                .claim(AuthConstants.EMAIL_ADDRESS, member.getEmailAddress())
                .claim(AuthConstants.EMAIL_VERIFIED, member.isEmailVerified())
                .issuedAt(new Date())
                .expiration(new Date(System.currentTimeMillis() + (long) accessTokenValidMinute * 2 * 50))
                .signWith(key)
                .compact();
    }

    public String generateRefreshToken(Member member) {
        return Jwts.builder()
                .claim(AuthConstants.ID, member.getId())
                .issuedAt(new Date())
                .expiration(new Date(System.currentTimeMillis() + (long) refreshTokenValidHour * 60 * 60 * 1000))
                .signWith(key)
                .compact();
    }

    public String getFreshAccessToken(String refreshToken){
        Claims claims = Jwts.parser()
                .verifyWith(key)
                .build()
                .parseSignedClaims(refreshToken)
                .getPayload();

        Long id = ((Integer) claims.get(AuthConstants.ID)).longValue();
        Member member = memberRepository.findById(id)
                .orElseThrow();

        return generateAccessToken(member);
    }

    public Map<String, Object> getClaims(String accessToken) {
        if (accessToken == null || accessToken.isBlank()) {
            return null;
        }
        try {
            Claims claims = Jwts.parser()
                    .verifyWith(key)
                    .build()
                    .parseSignedClaims(accessToken)
                    .getPayload();
            return new HashMap<>(claims);
        }catch (UnsupportedJwtException | MalformedJwtException e){
            throw new JwtException(e.getLocalizedMessage());
        }
    }


    public String generateTempToken(TempUserBeforeSignUp tempUserBeforeSignUp){
        return Jwts.builder()
                .claim(AuthConstants.OAUTH2_ID, tempUserBeforeSignUp.getOAuth2Id())
                .issuedAt(new Date())
                .expiration(new Date(System.currentTimeMillis() + (long) accessTokenValidMinute * 60 * 1000))
                .signWith(key)
                .compact();
    }
}

File: backend/src/main/java/com/NBE3_4_2_Team4/global/security/jwt/JwtManager.kt
package com.NBE3_4_2_Team4.global.security.jwt

import com.NBE3_4_2_Team4.domain.member.member.entity.Member
import com.NBE3_4_2_Team4.domain.member.member.repository.MemberRepository
import com.NBE3_4_2_Team4.global.security.user.tempUserBeforeSignUp.TempUserBeforeSignUp
import com.NBE3_4_2_Team4.standard.constants.AuthConstants
import io.jsonwebtoken.JwtException
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.MalformedJwtException
import io.jsonwebtoken.UnsupportedJwtException
import io.jsonwebtoken.security.Keys
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import java.util.*
import javax.crypto.SecretKey

@Component
class JwtManager(
    @Value("\${custom.jwt.secretKey:key}") jwtSecretKey: String,
    @Value("\${custom.jwt.accessToken.validMinute:30}") val accessTokenValidMinute: Int,
    @Value("\${custom.jwt.refreshToken.validHour:24}") val refreshTokenValidHour: Int,
    private val memberRepository: MemberRepository
) {
    private final var key: SecretKey

    init {
        try {
            val keyBytes = Base64.getDecoder().decode(jwtSecretKey)
            this.key = Keys.hmacShaKeyFor(keyBytes)
        } catch (e: IllegalArgumentException) {
            throw RuntimeException("키 값은 Base64로 디코딩 가능한 값이어야 합니다. yml 관련 파일을 확인해보세요.")
        }
    }

    fun generateAccessToken(member: Member): String {
        return Jwts.builder()
            .claim(AuthConstants.ID, member.id)
            .claim(AuthConstants.USERNAME, member.username)
            .claim(AuthConstants.NICKNAME, member.nickname)
            .claim(AuthConstants.ROLE, member.role.name)
            .claim(AuthConstants.OAUTH2_PROVIDER, member.oAuth2Provider.name)
            .claim(AuthConstants.EMAIL_ADDRESS, member.emailAddress)
            .claim(AuthConstants.EMAIL_VERIFIED, member.emailVerified)
            .issuedAt(Date())
            .expiration(Date(System.currentTimeMillis() + accessTokenValidMinute.toLong() * 2 * 50))
            .signWith(key)
            .compact()
    }

    fun generateRefreshToken(member: Member): String {
        return Jwts.builder()
            .claim(AuthConstants.ID, member.id)
            .issuedAt(Date())
            .expiration(Date(System.currentTimeMillis() + refreshTokenValidHour.toLong() * 60 * 60 * 1000))
            .signWith(key)
            .compact()
    }

    fun getFreshAccessToken(refreshToken: String?): String {
        val claims = Jwts.parser()
            .verifyWith(key)
            .build()
            .parseSignedClaims(refreshToken)
            .payload

        val id = (claims[AuthConstants.ID] as Int?)!!.toLong()
        val member = memberRepository.findById(id)
            .orElseThrow()

        return generateAccessToken(member)
    }

    fun getClaims(accessToken: String?): Map<String, Any>? {
        if (accessToken.isNullOrBlank()) {
            return null
        }
        try {
            val claims = Jwts.parser()
                .verifyWith(key)
                .build()
                .parseSignedClaims(accessToken)
                .payload
            return HashMap(claims)
        } catch (e: UnsupportedJwtException) {
            throw JwtException(e.localizedMessage)
        } catch (e: MalformedJwtException) {
            throw JwtException(e.localizedMessage)
        }
    }


    fun generateTempToken(tempUserBeforeSignUp: TempUserBeforeSignUp): String {
        return Jwts.builder()
            .claim(AuthConstants.OAUTH2_ID, tempUserBeforeSignUp.oAuth2Id)
            .issuedAt(Date())
            .expiration(Date(System.currentTimeMillis() + accessTokenValidMinute.toLong() * 60 * 1000))
            .signWith(key)
            .compact()
    }
}
